pkg/media/protectionMetadata_test.go

```diff
@@ -0,0 +1,325 @@
+
+	subsamples := &Subsamples{
+		SubsampleCount:  3,
+		SubsamplesTotal: 90,
+		Subsamples: []Subsample{
+			{
+				ClearDataByteSize:     10,
+				ProtectedDataByteSize: 20,
+			}, {
+				ClearDataByteSize:     10,
+				ProtectedDataByteSize: 20,
+			}, {
+				ClearDataByteSize:     10,
+				ProtectedDataByteSize: 20,
+			},
+		},
+	}
+
+	bytes := subsamples.Marshal()
+	require.Equal(t, []byte{3, 0, 10, 0, 0, 0, 20, 0, 10, 0, 0, 0, 20, 0, 10, 0, 0, 0, 20}, bytes)
+
+	_subsamples := &Subsamples{}
+	require.Nil(t, _subsamples.Unmarshal(bytes), "failed to unmarshal subsamples")
+
+	verifySubsamples(t, subsamples, _subsamples)
+}
+
+func TestProtectionMetadata_MarshalPattern(t *testing.T) {
+
+	pattern := &Pattern{
+		CryptByteBlock: 2,
+		SkipByteBlock:  5,
+	}
+
+	bytes := pattern.Marshal()
+	require.Equal(t, []byte{37}, bytes)
+
+	_pattern := &Pattern{}
+	require.Nil(t, _pattern.Unmarshal(bytes), "failed to unmarshal Pattern")
+
+	verifyPattern(t, pattern, _pattern)
+}
+
+func TestProtectionMetadata_MarshalProtectionMeta_Empty(t *testing.T) {
+
+	protectionMeta := &ProtectionMeta{}
+
+	extensions := protectionMeta.Marshal()
+
+	require.Empty(t, extensions, "extensions length should be 0")
+}
+
+func TestProtectionMetadata_MarshalProtectionMeta_Simple(t *testing.T) {
+
+	protectionMeta := &ProtectionMeta{
+		Meta: &Meta{
+			EncScheme:   2,
+			IsProtected: true,
+			Offset:      2,
+			Kid:         [16]byte{42, 42, 107, 105, 100, 69, 120, 97, 109, 112, 108, 101, 58, 41, 42, 42},
+			Iv:          [16]byte{42, 42, 105, 118, 69, 120, 97, 109, 112, 108, 101, 58, 41, 42, 42},
+		},
+	}
+
+	extensions := protectionMeta.Marshal()
+
+	require.Equal(t, 1, len(extensions), "extensions length should be 1")
+	require.NotNil(t, extensions[7], "extensions id should be 7")
+	require.Equal(t, []byte{40, 2, 42, 42, 107, 105, 100, 69, 120, 97, 109, 112, 108, 101, 58, 41, 42, 42, 42, 42, 105, 118, 69, 120, 97, 109, 112, 108, 101, 58, 41, 42, 42, 0}, extensions[7])
+
+	_meta := &Meta{}
+	require.Nil(t, _meta.Unmarshal(extensions[7]), "failed to unmarshal meta from extensions")
+
+	verifyMeta(t, protectionMeta.Meta, _meta)
+}
+
+func TestProtectionMetadata_MarshalProtectionMeta_WithEmptySubsamples(t *testing.T) {
+
+	protectionMeta := &ProtectionMeta{
+		Meta: &Meta{
+			EncScheme:   2,
+			IsProtected: true,
+			Offset:      2,
+			Kid:         [16]byte{42, 42, 107, 105, 100, 69, 120, 97, 109, 112, 108, 101, 58, 41, 42, 42},
+			Iv:          [16]byte{42, 42, 105, 118, 69, 120, 97, 109, 112, 108, 101, 58, 41, 42, 42},
+		},
+		Subsamples: &Subsamples{
+			SubsampleCount: 0,
+		},
+	}
+
+	extensions := protectionMeta.Marshal()
+
+	require.Equal(t, 1, len(extensions), "extensions length should be 1")
+	require.NotNil(t, extensions[7], "extensions id should be 7")
+	require.Equal(t, []byte{40, 2, 42, 42, 107, 105, 100, 69, 120, 97, 109, 112, 108, 101, 58, 41, 42, 42, 42, 42, 105, 118, 69, 120, 97, 109, 112, 108, 101, 58, 41, 42, 42, 0}, extensions[7])
```

Comment:
The test should only ensure extension id 7 was being added.
Meta extension is already validated in it's own test, no need to duplicate the assertion again.
(Each extension is independent)